udffsck/udffsck.c

```diff
@@ -0,0 +1,3149 @@
+        exit(8); 
+    }
+#ifdef MEMTRACE
+    dbg("\tArea allocated, pointer: %p, offset 0x%x\n", ptr, offset);
+#else
+    dbg("\tArea allocated\n");
+#endif
+}
+
+/**
+ * \brief UDF VRS detection function
+ *
+ * This function is trying to find VRS at sector 16.
+ * It also do first attempt to guess sectorsize.
+ *
+ * \param[in] *dev memory mapped device array
+ * \param[in,out] *sectorsize found sectorsize candidate
+ * \param[in] force_sectorsize if -B param is used, this flag should be set
+ *                                and sectorsize should be used automatically.
+ * \return 0 -- UDF succesfully detected, sectorsize candidate found
+ * \return -1 -- found BOOT2 or CDW02. Unsupported for now
+ * \return 1 -- UDF not detected 
+ */
+int is_udf(int fd, uint8_t **dev, int *sectorsize, size_t st_size, int force_sectorsize) {
+    struct volStructDesc vsd;
+    struct beginningExtendedAreaDesc bea;
+    struct volStructDesc nsr;
+    struct terminatingExtendedAreaDesc tea;
+    int ssize = 512;
+    int notFound = 0;
+    int foundBEA = 0;
+    uint32_t chunk = 0;
+    uint32_t chunksize = CHUNK_SIZE;
+
+    //void map_chunk(int fd, uint8_t **dev, uint32_t chunk, uint64_t st_size) {
+
+    for(int it=0; it<5; it++, ssize *= 2) {
+        if(force_sectorsize) {
+            ssize = *sectorsize;
+            it = INT_MAX - 1; //End after this iteration
+            dbg("Forced sectorsize\n");
+        }
+
+        dbg("Try sectorsize %d\n", ssize);
+
+        for(int i = 0; i<6; i++) {
+            chunk = (16*BLOCK_SIZE+i*ssize)/chunksize; 
+            map_chunk(fd, dev, chunk, st_size, __FILE__, __LINE__);
+            dbg("try #%d at address 0x%x, chunk %d, chunk address: 0x%x\n", i, 16*BLOCK_SIZE+i*ssize, chunk, (16*BLOCK_SIZE+i*ssize)%chunksize);
+#ifdef MEMTRACE
+            dbg("Chunk pointer: %p\n", dev[chunk]);
+#endif
+            memcpy(&vsd, dev[chunk]+(16*BLOCK_SIZE+i*ssize)%chunksize, sizeof(vsd));
+            dbg("vsd: type:%d, id:%s, v:%d\n", vsd.structType, vsd.stdIdent, vsd.structVersion);
+
+            if(!strncmp((char *)vsd.stdIdent, VSD_STD_ID_BEA01, 5)) {
+                //It's Extended area descriptor, so it might be UDF, check next sector
+                memcpy(&bea, &vsd, sizeof(bea)); // store it for later
+                foundBEA = 1; 
+            } else if(!strncmp((char *)vsd.stdIdent, VSD_STD_ID_BOOT2, 5)) {
+                err("BOOT2 found, unsuported for now.\n");
+                unmap_chunk(dev, chunk, st_size);
+                return -1;
+            } else if(!strncmp((char *)vsd.stdIdent, VSD_STD_ID_CD001, 5)) { 
+                //CD001 means there is ISO9660, we try search for UDF at sector 18
+            } else if(!strncmp((char *)vsd.stdIdent, VSD_STD_ID_CDW02, 5)) {
+                err("CDW02 found, unsuported for now.\n");
+                unmap_chunk(dev, chunk, st_size);
+                return -1;
+            } else if(!strncmp((char *)vsd.stdIdent, VSD_STD_ID_NSR01, 5)) {
```

Comment:
NSR01 is IIRC not allowed by any UDF revision. Have you ever seen a UDF disk with NSR01?

Reply:
True. Removed in 8aeeb8f